### PR TITLE
Allow to specify target versions of ECMAScript > 2015.

### PR DESCRIPTION
**Pull request by sjrd**
We introduce a new setting ESFeatures.esVersion to select specific ES versions to target for compliance. Values of `ESVersion` are defined from ES 5.1 up to ES 2020. The existing `useECMAScript2015` is remapped to be the same as `esVersion >= ESVersion.ES2015`.

We leverage `esVersion >= ESVersion.ES2017` in one place in the core JS lib. More tests like that could be introduced in the future.

We add `esVersion` in the generated `BuildInfo` for the test suite, and use it to better constrain some our tests of JS features.

The `esVersion` is also made available to user-space libraries through `LinkingInfo.assumedESVersion`, like `assumeES6`. This is not used yet in this commit, but will be used by the new `regex` support.

We enhance `NodeJSEnvForcePolyfills` and our Jenkins scripts to test with various target versions of ECMAScript.

---

Extracted from #4455 because I think this is generally useful, and it can be independently reviewed.

### CODE REVIEW

File: library/src/main/scala/scala/scalajs/js/BigInt.scala [hunk 1]
```diff
@@ -73,7 +73,7 @@ final class BigInt private[this] () extends js.Object {
 }
 
 
-/** <span class="badge badge-ecma2019" style="float: right;">ECMAScript 2019</span>
+/** <span class="badge badge-ecma2020" style="float: right;">ECMAScript 2020</span>
```
Comment (gzm0) [inline library/src/main/scala/scala/scalajs/js/BigInt.scala:76]: Put these into a separate commit? (IIUC these are essentially doc fixes discovered by these changes).

File: library/src/main/scala/scala/scalajs/runtime/LinkingInfo.scala [hunk 2]
```diff
@@ -16,7 +16,22 @@ import scala.scalajs.js
...
-  /** Whether we are assuming ECMAScript 6 support or not. */
+  /** Assumed version of ECMAScript.
+   *
+   *  This is an integer that represents the *edition* of the ECMAScript
+   *  Language Specification. For example, ECMAScript 2015 is represented with
+   *  the value `6`.
+   *
+   *  As an exception, ECMAScript 5.1 is represented with the value `5`.
+   *
+   *  If it is undefined, the value should be inferred from `assumingES6`.
```
Comment (gzm0) [inline library/src/main/scala/scala/scalajs/runtime/LinkingInfo.scala:27]: Put the more verbose description on `library.LinkingInfo`?

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala [hunk 3]
```diff
@@ -65,7 +78,7 @@ final class ESFeatures private (
    *  - In Script (`NoModule`) mode, top-level exports are defined as `let`s
    *    rather than `var`s, and behave as such.
    */
-  val useECMAScript2015 = _useECMAScript2015
+  val useECMAScript2015 = esVersion >= ESVersion.ES2015
```
Comment (gzm0) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala:81]: Shall we deprecate this?

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala [hunk 4]
```diff
@@ -164,6 +189,7 @@ final class ESFeatures private (
 
   override def toString(): String = {
     s"""ESFeatures(
+       |  esVersion = $esVersion,
        |  useECMAScript2015 = $useECMAScript2015,
```
Comment (gzm0) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala:193]: Remove this?

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala [hunk 5]
```diff
@@ -44,17 +44,30 @@ final class ESFeatures private (
...
+        _esVersion = ESVersion.ES2015,
         _allowBigIntsForLongs = false,
         _avoidClasses = true,
         _avoidLetsAndsConsts = true
     )
   }
 
+  /** The ECMAScript version that is assumed to be supported by the runtime.
+   *
+   *  Default: `ESVersion.ES2015`
```
Comment (gzm0) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala:56]: The documentation of the `Defaults` value on the companion also need this.

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala [hunk 6]
```diff
@@ -164,6 +189,7 @@ final class ESFeatures private (
 
   override def toString(): String = {
     s"""ESFeatures(
+       |  esVersion = $esVersion,
```
Comment (gzm0) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala:192]: The fingerprint also needs to be updated (I really don't like that we need to do all this manually, it's just too easy to forget :().

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESVersion.scala [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+import Fingerprint.FingerprintBuilder
+
+final class ESVersion private (val edition: Int, val name: String)
+    extends Ordered[ESVersion] {
+
+  import ESVersion._
+
+  def compare(that: ESVersion): Int = this.edition.compareTo(that.edition)
+
+  override def equals(that: Any): Boolean = that match {
```
Comment (gzm0) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESVersion.scala:24]: `this eq that`?

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESVersion.scala [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    case that: ESVersion => this.edition == that.edition
+    case _               => false
+  }
+
+  override def hashCode(): Int = {
```
Comment (gzm0) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESVersion.scala:29]: Simply `edition`?

File: project/Build.scala [hunk 9]
```diff
@@ -2006,6 +2008,7 @@ object Build {
               "compliantModuleInit" -> (sems.moduleInit == CheckedBehavior.Compliant),
               "strictFloats" -> sems.strictFloats,
               "productionMode" -> sems.productionMode,
+              "esVersion" -> linkerConfig.esFeatures.esVersion.edition,
               "es2015" -> linkerConfig.esFeatures.useECMAScript2015
```
Comment (gzm0) [inline project/Build.scala:2012]: Remove this?

File: test-suite/js/src/main/scala-ide-stubs/org/scalajs/testsuite/utils/BuildInfo.scala [hunk 10]
```diff
@@ -28,5 +28,6 @@ private[utils] object BuildInfo {
   final val compliantModuleInit = false
   final val strictFloats = false
   final val productionMode = false
+  final val esVersion = 0
   final val es2015 = false
```
Comment (gzm0) [inline test-suite/js/src/main/scala-ide-stubs/org/scalajs/testsuite/utils/BuildInfo.scala:32]: Remove?

File: test-suite/js/src/main/scala/org/scalajs/testsuite/utils/Platform.scala [hunk 11]
```diff
@@ -41,10 +42,14 @@ object Platform {
...
-    js.typeOf(js.Dynamic.global.BigInt) != "undefined"
+    assumedESVersion >= ESVersion.ES2020 || js.typeOf(js.Dynamic.global.BigInt) != "undefined"
+
+  lazy val jsRegExps2018: Boolean =
+    assumedESVersion >= ESVersion.ES2018 || regexFeatureTest("(?<=a)(?<!b)\\p{L}\\P{L}", "us")
 
   def sourceMaps: Boolean = BuildInfo.hasSourceMaps && executingInNodeJS
 
+  def assumedESVersion: Int = BuildInfo.esVersion
   def assumeES2015: Boolean = BuildInfo.es2015
```
Comment (gzm0) [inline test-suite/js/src/main/scala/org/scalajs/testsuite/utils/Platform.scala:53]: Remove?

File: test-suite/js/src/test/scala/org/scalajs/testsuite/library/LinkingInfoTest.scala [hunk 12]
```diff
@@ -26,6 +27,19 @@ class LinkingInfoTest {
...
+
+  @Test def esVersionConstants(): Unit = {
+    assertEquals(5, ESVersion.ES5_1)
+    assertEquals(6, ESVersion.ES2015)
+    assertEquals(7, ESVersion.ES2016)
+    assertEquals(8, ESVersion.ES2017)
+    assertEquals(9, ESVersion.ES2018)
+    assertEquals(10, ESVersion.ES2019)
+    assertEquals(11, ESVersion.ES2020)
+  }
```
Comment (gzm0) [inline test-suite/js/src/test/scala/org/scalajs/testsuite/library/LinkingInfoTest.scala:44]: What is the point of this test? It just tests the definitions. IMO a more useful test would be that the values in linker and the one in library agree. But `assumedESVersion` does this implicitly (if we run it for all versions).

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala [hunk 13]
```diff
@@ -130,8 +143,20 @@ final class ESFeatures private (
...
+   *
+   *  If `false`, this method sets the `esVersion` to `ESVersion.ES5_1`.
+   *  Otherwise, if `esVersion` was below `ES2015`, it sets it to `ES2015`.
+   *  Otherwise, it returns the same configuration of `ESFeatures`.
+   */
+  def withUseECMAScript2015(useECMAScript2015: Boolean): ESFeatures = {
+    if (!useECMAScript2015) withESVersion(ESVersion.ES5_1)
+    else if (esVersion == ESVersion.ES5_1) withESVersion(ESVersion.ES2015)
+    else this
+  }
```
Comment (gzm0) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala:159]: Add a unit test for this?

**Review by gzm0 — CHANGES_REQUESTED** [review]

**Comment by gzm0** [comment]
Hahaha, there is also a typo in the title EMCA vs ECMA

**Comment by sjrd** [comment]
I think we should keep `useECMAScript2015` as not deprecated, along with all its friends. IMO there is a qualitative difference between ES2015 and the following ones in Scala.js. Our semantics are specified in terms of ES2015, and when compiling to ES5.1, we do a best effort implementation of our semantics on top of that, but we lose some language semantics. That's why we've had this distinction for ages, even if we didn't have the following ECMAScript versions. Today we're adding them but mostly for library usage, not for language semantics. So I think the difference ES5.1 -> ES2015 still has significantly more weight than any of the following versions.

Does that make sense?

**Comment by gzm0** [comment]
> Does that make sense?

Yes it does. My concern here is two fold:
- User level (`ESLevel` / `LinkingInfo`): I'd rather not give two ways to do the same thing without at least a recommendation which one to use.
- Internally (`BuildInfo`): I'd rather not store data / have fields where some entirely derive of the others (`ESFeatures` already follows this).

Maybe we should remove storage and deprecate setters, but leave getters unchanged. (in any case, the qualitative differences of ES2015 should be made clear in the documentation on relevant fields on `ESFeatures`).

I think with this, we can argue that usage is less ambiguous:
- setting: `withESVersion` shall be used.
- reading: `esVersion` shall be used if a library feature is needed, `useECMAScript2015` if a specific semantic only supported by ES6 is needed.

Nevertheless, ES6 keeps its special status, made clear by the documentation explaining when and why `useECMAScript2015` shall be used over `esVersion >= ESVersions.ES2015`.

WDYT?

**Comment by sjrd** [comment]
That sounds very good, yes. Thank you. I'll update that on Tuesday, after the Easter break.

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala (hunk 3 above)
Comment (sjrd) [reply linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala:84]: Documented as discussed.

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala (hunk 4 above)
Comment (sjrd) [reply linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala:199]: Keeping it, given the discussion.

File: test-suite/js/src/main/scala/org/scalajs/testsuite/utils/Platform.scala (hunk 11 above)
Comment (sjrd) [reply test-suite/js/src/main/scala/org/scalajs/testsuite/utils/Platform.scala]: Kept and documented as discussed.

File: test-suite/js/src/test/scala/org/scalajs/testsuite/library/LinkingInfoTest.scala (hunk 12 above)
Comment (sjrd) [reply test-suite/js/src/test/scala/org/scalajs/testsuite/library/LinkingInfoTest.scala:45]: Since the numeric values are meaningful on their own, I thought they should be tested. I added a comment saying that.

**Comment by sjrd** [comment]
Updated.

File: test-suite/js/src/main/scala/org/scalajs/testsuite/utils/Platform.scala [hunk 14]
```diff
@@ -32,20 +33,37 @@ object Platform {
...
+  lazy val jsRegExps2018: Boolean =
+    assumedESVersion >= ESVersion.ES2018 || regexFeatureTest("(?<=a)(?<!b)\\p{L}\\P{L}", "us")
 
   def sourceMaps: Boolean = BuildInfo.hasSourceMaps && executingInNodeJS
 
-  def assumeES2015: Boolean = BuildInfo.es2015
+  /** The assumed ECMAScript version.
+   *
+   *  Prefer this property over `assumeES2015` when testing for support of
+   *  library functions.
```
Comment (gzm0) [inline test-suite/js/src/main/scala/org/scalajs/testsuite/utils/Platform.scala:55]: I'm not sure this advice actually applies here: It seems the pattern emerges to do JS library feature tests in this file. So maybe we should say don't use this (except in this file and for the `LinkingInfoTest`)?

**Review by gzm0 — CHANGES_REQUESTED** [review]

File: test-suite/js/src/main/scala/org/scalajs/testsuite/utils/Platform.scala (hunk 14 above)
Comment (sjrd) [reply test-suite/js/src/main/scala/org/scalajs/testsuite/utils/Platform.scala:55]: Some library feature tests that are repeatedly used in the codebase are done here. But there are a few additional ones that are done on a case-by-case basis. For examples, the tests in `ObjectTest`.

**Comment by sjrd** [comment]
Looking once more into this, I figured we should go further than what we discussed previously. There was no reason to actually link together the ES version and the ES2015 semantics. So I added a separate `useECMAScript2015Semantics` configuration, and cleaned up everything on the way.

Now both are stored, because it is possible to use any `esVersion` with `useECMAScript2015Semantics == false`. The only forbidden combo is `esVersion == ES5_1` with `useECMAScript2015Semantics = true`.

I added a few preliminary commits which should be used regardless, then added two commits with those changes. If that's OK with you, I'll squash the last 3 commits.

**Comment by gzm0** [comment]
> Now both are stored, because it is possible to use any esVersion with useECMAScript2015Semantics == false. The only forbidden combo is esVersion == ES5_1 with useECMAScript2015Semantics = true.

Two main questions I have:
- Are we confident that we will not use semantics of a future ES standard for something existing in Scala.js? What do we do if that happens anyways? (just define `useECMAScript20XXSemantics`?).
- What is a use-case for targeting a higher library level, with lower semantic level? Would such use cases be better met with allowing to switch individual semantics on/off? (TBH, I'm having trouble imagining a use case for this).

FWIW: While writing this, I realized that the answer to the first question is "No". If we decide to settle to use big ints for longs, we have exactly that situation.

**Comment by sjrd** [comment]
> Are we confident that we will not use semantics of a future ES standard for something existing in Scala.js? What do we do if that happens anyways? (just define `useECMAScript20XXSemantics`?).
> 
> FWIW: While writing this, I realized that the answer to the first question is "No". If we decide to settle to use big ints for longs, we have exactly that situation.

In general, this would happen if:

* there are new features of ECMAScript that have new, unique semantics
* and we decide that the most appropriate way to express those new ES semantics is using existing Scala concepts.

This is what happened to `class`es and non-native JS classes, or to `=>` functions and `js.FunctionN`s.

The case of `BigInt`s for Longs is a bit special. In that case, the ES feature choice is not so much dictated by semantics than by (supposed) performance gains. We would never do that for the semantics aspects, since a Long is in fact not adequate to represent the semantics of BigInts. Now that I think about it, I'm wondering whether we should *box* the BigInts in a `$Long` when it escapes to a JS context (or a generic context), in order to hide the semantics mismatch. But that's another story.

> What is a use-case for targeting a higher library level, with lower semantic level? Would such use cases be better met with allowing to switch individual semantics on/off? (TBH, I'm having trouble imagining a use case for this).

IMO the typical use case would be: I have a codebase that works with the current semantics (say ES2015 semantics); it wouldn't work with the newer semantics (say ES2021), yet I'm targeting an environment know to support ES2021. I'd like to take advantage of all the performance and/or code size improvements I can get from ES2021, without compromising the semantics of my codebase today. Sure, it's niche; perhaps even contrived. But it seems legit enough to me.

Besides the use case, though, I came to this from the point of view of the linker code. It was clearly use `useECMAScript2015` for two entirely different purposes. Actually decoupling them seems to produce cleaner code in the linker, IMO.

**Comment by gzm0** [comment]
> Actually decoupling them seems to produce cleaner code in the linker, IMO.

In that case (I'll review from this POV in detail), we should definitely do it internally.

> Sure, it's niche; perhaps even contrived. But it seems legit enough to me.

I'm a bit concerned about the additional complexity that this introduces in the linker configuration. I'd like to avoid adding a configuration option that isn't actually needed by anyone (both from a user and a maintenance POV).

**Comment by sjrd** [comment]
> > Sure, it's niche; perhaps even contrived. But it seems legit enough to me.
> 
> I'm a bit concerned about the additional complexity that this introduces in the linker configuration. I'd like to avoid adding a configuration option that isn't actually needed by anyone (both from a user and a maintenance POV).

That is true. It's also why I kept the last commit separate for now. If we don't want the additional complexity, but we want the internal cleanup of separating the two concepts, we can remove the last commit ("And now we can separately configure them").

**Comment by gzm0** [comment]
After looking at the code in more detail, I think we should not make `useECMAScript2015Semantics` configurable for now but:
- Definitely make the distinction internally.
- Make it clear in the documentation that `useECMAScript2015Semantics` might become configurable in the future.

If that's OK with you, I'll fully review once you squashed the commits.

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala [hunk 15]
```diff
@@ -35,7 +35,8 @@ final class ESFeatures private (
     /* We define `val`s separately below so that we can attach Scaladoc to them
      * (putting Scaladoc comments on constructor param `val`s has no effect).
      */
-    _useECMAScript2015: Boolean,
+    _esVersion: ESVersion,
```
Comment (gzm0) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ESFeatures.scala:38]: The doc above states semantics about certain prefixes in here. We should probably revise the doc or stay consistent with the prefixes.

**Comment by sjrd** [comment]
Updated:

* Removed the last commit
* Squashed
* Addressed comments and revisited a few more Scaladoc comments
* Adapted the commit message.

File: library/src/main/scala/scala/scalajs/runtime/LinkingInfo.scala [hunk 16]
```diff
@@ -16,7 +16,23 @@ import scala.scalajs.js
...
+   *
+   *  As an exception, ECMAScript 5.1 is represented with the value `5`.
+   */
+  val esVersion: Int
+
+  /** Whether Scala.js language features use ECMAScript 2015 (edition 6)
+   *  semantics or not.
+   *
+   *  For historical reasons, this is called `assumingES6`, but a better name
+   *  would be `useECMAScript2015Semantics`.
```
Comment (gzm0) [inline library/src/main/scala/scala/scalajs/runtime/LinkingInfo.scala:34]: Just to check my understanding: We cannot re-name this, otherwise a new linker will not properly link an old library anymore, right?

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala [hunk 17]
```diff
@@ -2696,8 +2699,8 @@ private[emitter] class FunctionEmitter(sjsGen: SJSGen) {
           js.New(transformExprNoChar(constr), args.map(transformJSArg))
 
         case Transient(JSNewVararg(constr, argsArray)) =>
-          assert(!esFeatures.useECMAScript2015,
-              s"generated a JSNewVargs in ES 2015 mode at ${tree.pos}")
+          assert(!es2015,
+              s"generated a JSNewVargs with ES 2015+ at ${tree.pos}")
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala:2703]: remove the line break?

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/JSGen.scala [hunk 18]
```diff
@@ -116,7 +116,7 @@ private[emitter] final class JSGen(val config: Emitter.Config) {
 
   def genArrowFunction(args: List[ParamDef], restParam: Option[ParamDef], body: Tree)(
       implicit pos: Position): Function = {
-    Function(useArrowFunctions, args, restParam, body)
+    Function(esFeatures.esVersion >= ESVersion.ES2015, args, restParam, body)
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/JSGen.scala:119]: Hah, this is subtle. Might deserve a comment :)

File: test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/FunctionTest.scala [hunk 19]
```diff
@@ -47,7 +47,7 @@ class FunctionTest {
   }
 
   @Test def functionWithConversionIsAnArrowFunction(): Unit = {
-    assumeTrue("In ES 5.1, arrow functions do not exist", assumingES6)
+    assumeTrue("In ES 5.1, arrow functions do not exist", useECMAScript2015Semantics)
```
Comment (gzm0) [inline test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/FunctionTest.scala:50]:
```
assumeTrue("Requires ECMAScript 2015 semantics", useECMAScript2015Semantics)
```

also below?

File: test-suite/js/src/test/scala/org/scalajs/testsuite/library/WrappedSetTest.scala [hunk 20]
```diff
@@ -17,12 +17,16 @@ import org.junit.Assume.assumeTrue
...
+import scala.scalajs.js
+
+import org.scalajs.testsuite.utils.Platform._
 
 object WrappedSetTest {
   @BeforeClass
   def assumeRuntimeSupportsSet(): Unit = {
-    assumeTrue("Assume ES6", LinkingInfo.assumingES6)
+    assumeTrue("Requires js.Set support",
+        assumeES2015 || js.typeOf(js.Dynamic.global.Set) != "undefined")
```
Comment (gzm0) [inline test-suite/js/src/test/scala/org/scalajs/testsuite/library/WrappedSetTest.scala:29]: We should factor this out (in another PR).

**Review by gzm0 — APPROVED** [review]

File: library/src/main/scala/scala/scalajs/runtime/LinkingInfo.scala (hunk 16 above)
Comment (sjrd) [reply library/src/main/scala/scala/scalajs/runtime/LinkingInfo.scala:34]: Yes, that's right.

**Comment by sjrd** [comment]
Updated to address the latest review.
